### ISSUE TITLE
Notebook manager doesn't display properly when redirected to an "Output" widget

### ISSUE THREAD

**Issue by cubano21** [Rockhopper-Technologies/enlighten#60]
**Bug Description**
When using enlighten in a JupyterLab notebook, the progress bar shows up normally in the regular cell output, but when captured and redirected into an "Output" widget from ipywidgets, the bar does not update properly. Regular stdout is captured normally, but only the initial creation of the progress bar appears. Inspecting the manager shows that everything is computed as expected, only the display is affected.

**To Reproduce**
Here is my example notebook code to reproduce:
cell 1:
```from IPython.display import display, HTML, clear_output
import ipywidgets as widgets
_output = widgets.Output()
_clear = widgets.Button(description='Clear', disabled=False, button_style='', icon='fire-extinguisher')
def clear(*args):
    with _output:
        clear_output()
_clear.on_click(clear)
widgets.VBox(_clear, _output])
```

cell2:
```
import time
import enlighten
pbar_style=' ⡀⡄⡆⡇⣇⣧⣷⣿█'
with _output:
    manager = enlighten.get_manager()
    status_bar = manager.status_bar('Initializing', justify=enlighten.Justify.CENTER)
    barfmt = '{desc:>{maxlen}}{desc_pad}{percentage:3.0f}%|{bar}| {count:{lencnt}d}/{total:{lencnt}d} [{elapsed}<{eta}, {interval:5.3H}s/{unit}]'
    pbar = manager.counter(total=10, desc=f'test bar:', unit='run', bar_format=barfmt, maxlen=8, lencnt=4,
                                            series=pbar_style)
    
    for i in range(10):
        time.sleep(0.1)
        print(i)
        pbar.update()
```

Which produces the output:
![image](https://github.com/Rockhopper-Technologies/enlighten/assets/143648149/6065d7e3-ed9f-49d1-84d8-9b661abdfa53)

**Workaround / Proposed Fix**
I was able to workout this issue by essentially changing the _display field from "DisplayHandle()" to an "HTML" widget and then updating the value of the widget when flushing the streams. Below is my code to monkey-patch enlighten (created as a function in case there is a need to not patch):
```
import ipywidgets as widgets
from IPython.display import display
def patch_enlighten():
    def _flush_streams(self):
        """
        Display buffered output
        """

        if not self.enabled:
            return

        self._html.data = '%s<div class="enlighten">\n%s\n</div>\n' % (
                          self._converter.style, '\n'.join(reversed(self._output)))

        if not self._primed:
            self._primed = True
            self._display = widgets.HTML()
            display(self._display)
        
        self._display.value = self._html.data
    import enlighten
    enlighten.NotebookManager._flush_streams = _flush_streams
```

After which, the progress bar appears and updates as expected in either the standard cell output or redirected through an "Output" widget:
![image](https://github.com/Rockhopper-Technologies/enlighten/assets/143648149/3a6b24d6-b48e-4f20-9929-136e4cdd0836)

**Environment (please complete the following information):**
 - Enlighten Version: 1.11.2

- Selected Jupyter core packages...
  - IPython          : 8.14.0
  - ipykernel        : 6.25.0
  - ipywidgets       : 8.1.0
  - jupyter_client   : 8.3.0
  - jupyter_core     : 5.3.1
  - jupyter_server   : 2.7.0
  - jupyterlab       : 4.0.4
  - nbclient         : 0.8.0
  - nbconvert        : 7.7.3
  - nbformat         : 5.9.2

**Comment by avylove** [comment]
I can see why it would be desirable for this to work. However, there are two issues with your solution.

The first is it doesn't work if someone wants to rerun the block without using widgets.
The second is it would require all users to install ipywidgets.

I'd like to accommodate this use case, but I think it will require being able to detect this condition. I posed the question [here](https://discourse.jupyter.org/t/is-there-a-way-to-detect-when-running-in-a-with-ipywidgets-output-block/21188l).

**Comment by cubano21** [comment]
Fair point, and part of why I put my patch in a module function so that I can just patch it when using notebooks with widgets and leave it unchanged when running in the terminal.

It's possible that there is some other way of updating the content of the Output widget without using a nested widget, but I haven't gotten familiar enough with the internals of Jupyter notebooks or widgets to do so yet.

Perhaps it would be sufficient to attempt to import ipywidgets and catch the import error to decide which method to use (maybe a separate "WidgetManager" class served up by get_manager() to avoid a bunch of conditionals keeping track of which approach is in use). Not the most elegant solution, but if you are in a notebook environment and ipywidgets is available, the patched method seems to work fine in the standard cell output, and otherwise you couldn't be redirecting to an Output widget anyway.

**Comment by avylove** [comment]
Checking to see if ipywidgets is available wouldn't be sufficient. Essentially we'll want to know if it's within a `with` block. It should be possible to determine since it clearly results in different behavior. I just don't know those libraries, so it will require research or input from someone who does.

The behavior in your example is not the same. Try running the cell multiple times. The widgets do not seem to redraw where the  the standard display handles do.

**Comment by avylove** [comment]
I've looked into this a few times and not seeing an easy way to support this. Creating a widget-style manager would be relatively easy, but the question would be how to invoke it. The options are:

- Manually selected through dedicated manager or option
  - Not ideal, but potentially workable
- Detect when called within an Output widget
  - Ideal, but not sure how to do it
- Some other way to assume it's intended
  - Not sure what this would be

Alternatively, if there was a way to update the output, that would suffice. The problem is we update ``IPython.display.HTML.data`` and it seems that is parsed and displayed without retaining a handle.

I would like your scenario to work, but I don't know enough about how ipywidgets works internally to be able do detect it and got zero replies on the Jupyer Discourse. If someone can provide the needed info, I'm happy to look at this again.
